File: py/1100.py
from collections import deque
from sys import stdin, stdout


letters = [None, 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']
dx = [2, 2, -2, -2, 1, 1, -1, -1]
dy = [1, -1, 1, -1, 2, -2, 2, -2]


class Knight:
    def __init__(self, x, y, dist = 0) -> None:
        self.x = x
        self.y = y
        self.dist = dist


def check_bounds(x, y) -> bool:
    return x > 0 and x < 9 and y > 0 and y < 9


def solution(pos_x, pos_y, targ_x, targ_y) -> int:
    queue = deque()
    queue.append(Knight(pos_x, pos_y))
    visited = [[False for _ in range(9)] for _ in range(9)]
    visited[pos_x][pos_y] = True
    while queue:
        move: Knight = queue.popleft()
        if move.x == targ_x and move.y == targ_y:
            return move.dist
        for i in range(8):
            x = move.x + dx[i]
            y = move.y + dy[i]
            if check_bounds(x, y) and not visited[x][y]:
                visited[x][y] = True
                queue.append(Knight(x, y, move.dist + 1))


def main() -> None:
    input = stdin.readline
    print = stdout.write
    while True:
        try:
            fr, to = input().split()
            fr_a, fr_b = letters.index(fr[0]), int(fr[1])
            to_a, to_b = letters.index(to[0]), int(to[1])
            moves = solution(fr_a, fr_b, to_a, to_b)
            print(f'To get from {fr} to {to} takes {moves} knight moves.\n')
        except ValueError:
            break


if __name__ == '__main__':
    main()

File: cpp/1100.cpp
#include <iostream>
#include <string>
#include <queue>

using namespace std;

int dx[] = {2, 2, -2, -2, 1, 1, -1, -1};
int dy[] = {1, -1, 1, -1, 2, -2, 2, -2};

class Knight
{
    public:
        int x, y, dist;
        Knight(int a, int b, int c = 0)
        {
            x = a;
            y = b;
            dist = c;
        }
};

bool check_bounds(int x, int y)
{
    return (x > 0 && x < 9 && y > 0 && y < 9);
}

int solution(int pos_x, int pos_y, int targ_x, int targ_y)
{
    queue<Knight> queue;
    queue.push(Knight(pos_x, pos_y));
    bool visited[9][9] = {false};
    visited[pos_x][pos_y] = true;
    int i, x, y;
    while (!queue.empty()) {
        Knight move = queue.front();
        queue.pop();
        if (move.x == targ_x && move.y == targ_y) {
            return move.dist;
        }
        for (i = 0; i < 8; i++) {
            x = move.x + dx[i];
            y = move.y + dy[i];
            if (check_bounds(x, y) && !visited[x][y]) {
                visited[x][y] = true;
                queue.push(Knight(x, y, move.dist + 1));
            }
        }
    }
    return 0;
}

int main()
{
    string fr, to;
    int fr_x, fr_y, to_x, to_y, moves;
    while (cin >> fr >> to) {
        fr_x = fr[0] - 96;
        to_x = to[0] - 96;
        fr_y = fr[1] - '0';
        to_y = to[1] - '0';
        moves = solution(fr_x, fr_y, to_x, to_y);
        cout << "To get from " << fr << " to " << to << " takes " << moves << " knight moves.\n";
    }
    return 0;
}